### PR TITLE
Show recipe template kind and path in the status

### LINKED ISSUE #6440: Store and return the Recipe that was used as part of portable resources
## Overview

We should return `templateKind` and `templatePath` (and `templateVersion` if Terraform) as part of the portable resource body (see below).

Note that **this should be the templateKind and templatePath that were used when the portable resource was deployed**. If the environment is updated with a different path/kind then the portable resource should still return the old one from when the portable resource was deployed, and not the new one.

## Example

```bicep
import radius as rad

resource env 'Applications.Core/environments@2023-10-01-preview' = {
  name: 'default'
  properties: {
    compute: {
      kind: 'kubernetes'
      namespace: 'default'
    }
    recipes: {
      'Applications.Datastores/redisCaches': {
        'default': {
          templateKind: 'bicep'
          templatePath: 'radius.azurecr.io/recipes/local-dev/rediscaches'
        }
      }
    }
  }
}

resource app 'Applications.Core/applications@2023-10-01-preview' = {
  name: 'myapp'
}

resource redis 'Applications.Datastores/redisCaches@2023-10-01-preview' = {
  name: 'redis'
  properties: {
    environment: env.id
    application: app.id
  }
}
```

### Expected output

```bash
rad resource show rediscaches redis -o json
```

```diff
{
  "id": "/planes/radius/local/resourcegroups/default/providers/Applications.Datastores/redisCaches/redis",
  "location": "global",
  "name": "redis",
  "properties": {
    "application": "/planes/radius/local/resourceGroups/default/providers/applications.core/applications/demo",
    "environment": "/planes/radius/local/resourceGroups/default/providers/applications.core/environments/default",
    "host": "redis-hjo6ha3uqagio.default-demo.svc.cluster.local",
    "port": 6379,
    "provisioningState": "Succeeded",
    "recipe": {
      "name": "default"
    },
    "resourceProvisioning": "recipe",
    "status": {
+      "recipe": {
+        templateKind: "bicep",
+        templatePath: "radius.azurecr.io/recipes/local-dev/rediscaches:latest"
+      }
      "outputResources": [
        {
          "id": "/planes/kubernetes/local/namespaces/default-demo/providers/core/Service/redis-hjo6ha3uqagio",
          "radiusManaged": true
        },
        {
          "id": "/planes/kubernetes/local/namespaces/default-demo/providers/apps/Deployment/redis-hjo6ha3uqagio",
          "radiusManaged": true
        }
      ]
    },
    "tls": false,
    "username": ""
  },
  "systemData": {
    "createdAt": "0001-01-01T00:00:00Z",
    "createdBy": "",
    "createdByType": "",
    "lastModifiedAt": "0001-01-01T00:00:00Z",
    "lastModifiedBy": "",
    "lastModifiedByType": ""
  },
  "tags": {},
  "type": "Applications.Datastores/redisCaches"
}
```

[AB#9683](https://dev.azure.com/azure-octo/e61041b4-555f-47ae-95b2-4f8ab480ea57/_workitems/edit/9683)

### PR DESCRIPTION
**Pull request by vinayada1**
# Description

Add template kind and path to resource status. Then store these values and return in rad resource show

## Type of change

- This pull request adds or changes features of Radius and has an approved issue (issue link required).

Fixes: #6440 

### Summary
📝🚀🎁

This pull request adds support for template-based resource deployment in the `v20231001preview` API version of the radius project. It introduces the `RecipeData` type and the `Recipe` field in various resource status types, and adds custom JSON serialization and deserialization logic for them. It also updates the `ConvertFrom` functions and the test cases for the affected resource types to handle the new `Recipe` field.

> _`Recipe` field added_
> _To support template deployment_
> _A new feature blooms_

### Walkthrough
*  Add `RecipeData` type to represent the recipe data for a resource deployment using templates ([link](https://github.com/radius-project/radius/pull/6450/files?diff=unified&w=0#diff-21034b9d798fa570ecae4b7d54142c7f94d9d0f1b39c5f45f0c4c859674b19fbR1354-R1362), [link](https://github.com/radius-project/radius/pull/6450/files?diff=unified&w=0#diff-a50172617a56a90110cf640526438fbb2b1e3b46f1c22dc2f0119a1d513baedfR481-R489), [link](https://github.com/radius-project/radius/pull/6450/files?diff=unified&w=0#diff-4ba9624b522a455e82e79bb4fac573d7c8c8af0f39b9df75efa926034a9e6156R1103-R1133))
*  Add `Recipe` field to `ResourceStatus` type to hold the recipe data for a resource deployment ([link](https://github.com/radius-project/radius/pull/6450/files?diff=unified&w=0#diff-21034b9d798fa570ecae4b7d54142c7f94d9d0f1b39c5f45f0c4c859674b19fbR1455-R1457), [link](https://github.com/radius-project/radius/pull/6450/files?diff=unified&w=0#diff-a50172617a56a90110cf640526438fbb2b1e3b46f1c22dc2f0119a1d513baedfR527-R529), [link](https://github.com/radius-project/radius/pull/6450/files?diff=unified&w=0#diff-4ba9624b522a455e82e79bb4fac573d7c8c8af0f39b9df75efa926034a9e6156R1236), [link](https://github.com/radius-project/radius/pull/6450/files?diff=unified&w=0#diff-4ba9624b522a455e82e79bb4fac573d7c8c8af0f39b9df75efa926034a9e6156R1255-R1257))
*  Populate `Recipe` field in `ConvertFrom` functions of `Extender`, `DaprPubSubBroker`, `DaprSecretStore`, `DaprStateStore`, `MongoDatabase`, and `RedisCache` types ([link](https://github.com/radius-project/radius/pull/6450/files?diff=unified&w=0#diff-287413ad769e13989c3e7d75c20c1399ce90a2ccd7be0e1a57918968f5490729R90-R96), [link](https://github.com/radius-project/radius/pull/6450/files?diff=unified&w=0#diff-3a000e41539b8b6457f28fc3587853d2bc3aac0d9fac097f33cd622347835e1bR144-R150), [link](https://github.com/radius-project/radius/pull/6450/files?diff=unified&w=0#diff-eb720f089b10b32de7e0206eb26ba88405e39096da247e3a35d9d8cd125e38d8R136-R143), [link](https://github.com/radius-project/radius/pull/6450/files?diff=unified&w=0#diff-a8a8f5a7060d5275f77ceec374c048e6e38430ab02cfd60dffdcdc98020a4c8aR127-R133), [link](https://github.com/radius-project/radius/pull/6450/files?diff=unified&w=0#diff-a3deb9a38591d44ecf385106451d056ec8baeb200b3d558396b430952ecefcb2R111-R117), [link](https://github.com/radius-project/radius/pull/6450/files?diff=unified&w=0#diff-f3da6c5dde4eaacd441388811703b7cb0e650b6ae6ba9c9b6e1c9c6c6b580fb8R110-R116))
*  Update test cases for `ConvertFrom` functions of `Extender`, `DaprPubSubBroker`, `DaprSecretStore`, `DaprStateStore`, `MongoDatabase`, and `RedisCache` types to reflect the `Recipe` field ([link](https://github.com/radius-project/radius/pull/6450/files?diff=unified&w=0#diff-bd287311ae13f33257e57cc10affc0dfa7c68d06294186501f62d76e1a85faf3L207-R212), [link](https://github.com/radius-project/radius/pull/6450/files?diff=unified&w=0#diff-3e36042b3fa516b14d6e34fee3536dcef97fb5c73325a4bc0244d8f1576daed1L220-R225), [link](https://github.com/radius-project/radius/pull/6450/files?diff=unified&w=0#diff-ed4bd83d135ddb4290a01afbb8d22312a733f13dba286675c9d466d7408bb002L184-R189), [link](https://github.com/radius-project/radius/pull/6450/files?diff=unified&w=0#diff-0951811300b3dbd68d11e79138bdaec82cb3624aca2cc15d718c92b7b374b7dcR173), [link](https://github.com/radius-project/radius/pull/6450/files?diff=unified&w=0#diff-0951811300b3dbd68d11e79138bdaec82cb3624aca2cc15d718c92b7b374b7dcR179-R184), [link](https://github.com/radius-project/radius/pull/6450/files?diff=unified&w=0#diff-10f63831be836e2619070614aa768a9a6351f3a1f388942fbf5478d611a7a84bR299-R302), [link](https://github.com/radius-project/radius/pull/6450/files?diff=unified&w=0#diff-3c767c971732c83e96fc31b7e924cf66d4fb853fc9d42d83936283b895085a0dL224-R229))
*  Add sample data models for testing `ConvertFrom` functions of `Extender`, `DaprPubSubBroker`, `DaprSecretStore`, and `DaprStateStore` types with the `Recipe` field ([link](https://github.com/radius-project/radius/pull/6450/files?diff=unified&w=0#diff-e01ee096c196892a0b5be43c5e28fd01759162b492878771d860b7fef8464820L22-R26), [link](https://github.com/radius-project/radius/pull/6450/files?diff=unified&w=0#diff-105a4d9930462d0e04049cdad7cc1c97fa30bc4707c4d9e4a2fd4517e68a2d03L24-R28), [link](https://github.com/radius-project/radius/pull/6450/files?diff=unified&w=0#diff-2d3dd347fde4e6895a7c22d4ec1c665d56b30551ecb6e2193e2e11c4ce6e9a46L24-R28), [link](https://github.com/radius-project/radius/pull/6450/files?diff=unified&w=0#diff-11867dad48397245fdab0cc50c6eb4b99e47f7cb20dac8118a6d06af75bc7056L24-R28))
*  Simplify test case for `ConvertFrom` function of `DaprStateStore` type by removing redundant assignment of `Status` field ([link](https://github.com/radius-project/radius/pull/6450/files?diff=unified&w=0#diff-0951811300b3dbd68d11e79138bdaec82cb3624aca2cc15d718c92b7b374b7dcL158))

### CODE REVIEW

**Review by ytimocin — COMMENTED** [review]
We need **templateVersion** for the Terraform recipes.

**Review by kachawla — CHANGES_REQUESTED** [review]
Can we please build clarity on re-deployment scenario before merging this PR? https://github.com/radius-project/radius/pull/6450#discussion_r1355772650

File: pkg/datastoresrp/api/v20231001preview/datamodel_util_test.go
Comment (ytimocin) [inline pkg/datastoresrp/api/v20231001preview/datamodel_util_test.go:217]:
```suggestion
func Test_fromRecipeStatus(t *testing.T) {
```

File: pkg/datastoresrp/api/v20231001preview/testdata/mongodatabaseresourcedatamodel_recipe.json
Comment (ytimocin) [inline pkg/datastoresrp/api/v20231001preview/testdata/mongodatabaseresourcedatamodel_recipe.json:29]: Do we need examples with Terraform?

File: pkg/portableresources/processors/validator_test.go
Comment (ytimocin) [inline pkg/portableresources/processors/validator_test.go:776]: Should we have an example with a Recipe?

File: pkg/portableresources/types.go
Comment (ytimocin) [reply pkg/portableresources/types.go:62]: Are we not supposed to have **TemplateVersion** here?

File: pkg/recipes/types_test.go
Comment (ytimocin) [inline pkg/recipes/types_test.go:30]: Should we add some tests with a Recipe?

File: pkg/portableresources/types.go
Comment (ytimocin) [reply pkg/portableresources/types.go:62]: Or should these be deleted because we defined the RecipeStatus somewhere else?

File: pkg/recipes/types_test.go
Comment (vinayada1) [reply pkg/recipes/types_test.go:30]: No changes to PrepareRecipeResponse as it is called only sometimes. The changes are outside.

File: pkg/portableresources/types.go
Comment (vinayada1) [reply pkg/portableresources/types.go]: Removed

File: pkg/datastoresrp/api/v20231001preview/testdata/mongodatabaseresourcedatamodel_recipe.json
Comment (vinayada1) [reply pkg/datastoresrp/api/v20231001preview/testdata/mongodatabaseresourcedatamodel_recipe.json:29]: pkg/corerp/api/v20231001preview/testdata/environmentresource-terraformrecipe-localpath.json

File: pkg/portableresources/backend/controller/createorupdateresource.go
Comment (vinayada1) [reply pkg/portableresources/backend/controller/createorupdateresource.go]: I see what you are saying. I removed the nil check to go through this even during redeployment. We are storing the status on the resource. So if the env is updated, that wont affect the values returned as mentioned in the issue.

**Review by ytimocin — DISMISSED** [review]
Thanks for the change! LGTM.

File: pkg/portableresources/backend/controller/createorupdateresource.go
Comment (kachawla) [inline pkg/portableresources/backend/controller/createorupdateresource.go:129]: `recipeOutput.Status` should always be populated as long as a recipe is used for deployment, and ignoring this can cause bugs to slip in. So it would be better to error out if this happens, let me know what you think.

File: pkg/portableresources/backend/controller/createorupdateresource.go
Comment (kachawla) [inline pkg/portableresources/backend/controller/createorupdateresource.go:133]: Since this is optional, should a nil check be added?

File: pkg/datastoresrp/api/v20231001preview/mongodatabase_conversion_test.go
Comment (kachawla) [inline pkg/datastoresrp/api/v20231001preview/mongodatabase_conversion_test.go:301]: Should this use `MustPopulateResourceStatusWithRecipe`?

File: pkg/datastoresrp/api/v20231001preview/rediscache_conversion_test.go
Comment (kachawla) [inline pkg/datastoresrp/api/v20231001preview/rediscache_conversion_test.go:226]: Should this use `MustPopulateResourceStatusWithRecipe`?

File: pkg/datastoresrp/api/v20231001preview/sqldatabase_conversion_test.go
Comment (kachawla) [inline pkg/datastoresrp/api/v20231001preview/sqldatabase_conversion_test.go:198]: Same here - Should this use `MustPopulateResourceStatusWithRecipe`?

File: pkg/portableresources/backend/controller/createorupdateresource.go
Comment (kachawla) [inline pkg/portableresources/backend/controller/createorupdateresource.go:126]: Do we have unit tests for changes made in this file?

File: pkg/portableresources/backend/controller/createorupdateresource.go
Comment (vinayada1) [reply pkg/portableresources/backend/controller/createorupdateresource.go:133]: It's a string not a ptr...therefore did not add a nil check

File: pkg/portableresources/backend/controller/createorupdateresource.go
Comment (vinayada1) [reply pkg/portableresources/backend/controller/createorupdateresource.go:126]: It's a bit hard to test these with the mocks. That's why probably other fields updated by the code are not validated in the UT as well. This will be covered by the functional test though

File: pkg/datastoresrp/api/v20231001preview/mongodatabase_conversion_test.go
Comment (vinayada1) [inline pkg/datastoresrp/api/v20231001preview/mongodatabase_conversion_test.go:297]: Cannot use MustPopulateResourceStatusWithRecipe due to OutputResources: nil check

Comment (kachawla) [reply pkg/datastoresrp/api/v20231001preview/mongodatabase_conversion_test.go]: Not sure why the output resource was nil, but looks like we were able to replace it with MustPopulateResourceStatusWithRecipe, so resolving this.

File: pkg/daprrp/api/v20231001preview/secretstore_conversion_test.go
Comment (kachawla) [inline pkg/daprrp/api/v20231001preview/secretstore_conversion_test.go:189]:
```suggestion
					Status: resourcetypeutil.MustPopulateResourceStatusWithRecipe(&ResourceStatus{}),
```

File: pkg/messagingrp/api/v20231001preview/rabbitmq_conversion_test.go
Comment (kachawla) [inline pkg/messagingrp/api/v20231001preview/rabbitmq_conversion_test.go:197]:
```suggestion
					Status: resourcetypeutil.MustPopulateResourceStatusWithRecipe(&ResourceStatus{}),
```

File: pkg/portableresources/backend/controller/createorupdateresource.go
Comment (kachawla) [reply pkg/portableresources/backend/controller/createorupdateresource.go]: I think this code should be a part of validator.go where rest of the output processing happens as I mentioned earlier, that will solve the problem of unit testing it as well.

https://github.com/radius-project/radius/pull/6450#discussion_r1353851994

File: pkg/recipes/types.go
Comment (rynowak) [inline pkg/recipes/types.go:137]: If this is just going to set fields, do we need a function for it? I don't have a strong feeling about this, but IMO it's simpler if we just set the fields directly.

File: test/validation/shared.go
Comment (rynowak) [inline test/validation/shared.go:151]: Can you tell me a little bit more about what's going on here? It looks like `outputResources` might be either an array or map. That doesn't sound good if that's the case.

Comment (vinayada1) [reply test/validation/shared.go:151]: Earlier, the only fields in status were "outputResources" and therefore status looked like map with key = "outputresources" and value = resources and we were directly calling marshal on the status. Now, we have added recipe.
Screenshot 2023-10-30 at 4 06 50 PM

Therefore, we are now separating out status["outputResources"] and then marshaling and unmarshaling before validating the output resources.

File: pkg/recipes/driver/bicep.go
Comment (kachawla) [inline pkg/recipes/driver/bicep.go:368]: Do we need to add the `TemplateVersion` as empty or can we skip setting it here? It would be good to show template version field in the response only for scenarios where it makes sense.

File: test/functional/testUtil.go
Comment (kachawla) [inline test/functional/testUtil.go:104]: Were changes in this file intended or were meant for local testing?

File: pkg/datastoresrp/processors/mongodatabases/processor.go
Comment (kachawla) [inline pkg/datastoresrp/processors/mongodatabases/processor.go:35]: Could you help me understand why does it need to be `resource.ResourceMetadata().Status.Recipe` instead of `&resource.Properties.Status.Recipe`?

File: pkg/portableresources/backend/controller/createorupdateresource.go
Comment (kachawla) [inline pkg/portableresources/backend/controller/createorupdateresource.go:81]: Since this is already being set in the processors, why do we need this additional check here?

File: pkg/datastoresrp/api/v20231001preview/mongodatabase_conversion_test.go
Comment (kachawla) [inline pkg/datastoresrp/api/v20231001preview/mongodatabase_conversion_test.go:303]: Should this use `MustPopulateResourceStatusWithRecipe`?

**Review by kachawla — COMMENTED** [review]
Looks good overall, should be good to go after these last set of comments

File: test/functional/testUtil.go
Comment (vinayada1) [reply test/functional/testUtil.go:104]: Thanks for catching this!

File: pkg/datastoresrp/api/v20231001preview/mongodatabase_conversion_test.go
Comment (vinayada1) [reply pkg/datastoresrp/api/v20231001preview/mongodatabase_conversion_test.go:303]: In line 298, we want outputresources = nil. MustPopulateResourceStatusWithRecipe expects an output resource and so did not use it.

File: pkg/portableresources/backend/controller/createorupdateresource.go
Comment (vinayada1) [reply pkg/portableresources/backend/controller/createorupdateresource.go:81]: The processor sets the content of the pointer.
*v.Status = *output.Status

v.Status should not be nil here.

File: pkg/corerp/processors/extenders/processor.go
Comment (kachawla) [inline pkg/corerp/processors/extenders/processor.go:20]: Should this be passed by reference?

```suggestion
	validator := processors.NewValidator(&resource.ComputedValues, &resource.SecretValues, &resource.Properties.Status.OutputResources, &resource.Properties.Status.Recipe)
```

Comment (vinayada1) [reply pkg/corerp/processors/extenders/processor.go:20]: As discussed offline, this is a pointer. Will keep as is

**Review by kachawla — DISMISSED** [review]

File: typespec/radius/v1/resources.tsp
Comment (shalabhms) [inline typespec/radius/v1/resources.tsp:110]: consistent with other param descriptions.
```suggestion
  @doc("TemplatePath is the path of the recipe consumed by the portable resource upon deployment.")
```

**Review by shalabhms — DISMISSED** [review]

**Review by kachawla — DISMISSED** [review]

**Review by shalabhms — APPROVED** [review]

**Review by kachawla — COMMENTED** [review]
@vinayada1 Can you merge the bicep changes as well https://github.com/radius-project/bicep/pull/684?

**Pull request merged**
